File: scripts/hot_reply_watcher.py
#!/usr/bin/env python3
"""Instantly hot-reply watcher — fires every 2 hours during UK working hours."""

import json
import os
import sys
import time
from datetime import datetime, timedelta, timezone
from html import escape

import anthropic
import requests

INSTANTLY_BASE = "https://api.instantly.ai/api/v2"
TELEGRAM_BASE = "https://api.telegram.org"
CAL_LINK = "https://cal.com/voqalai/discovery"
HOT_CLASSES = {"INTERESTED", "BOOK_A_CALL", "OBJECTION"}
ALL_CLASSES = {
    "INTERESTED", "BOOK_A_CALL", "OBJECTION", "OOO",
    "UNSUBSCRIBE", "NOT_INTERESTED", "WRONG_PERSON", "QUESTION",
}


def get_env(name: str) -> str:
    val = os.environ.get(name, "")
    if not val:
        raise RuntimeError(f"Missing required env var: {name}")
    return val


def fetch_replies(campaign_id: str, since_iso: str, api_key: str) -> list[dict]:
    headers = {"Authorization": f"Bearer {api_key}"}
    params = {"campaign_id": campaign_id, "limit": 100}
    resp = requests.get(
        f"{INSTANTLY_BASE}/emails/reply",
        headers=headers,
        params=params,
        timeout=15,
    )
    resp.raise_for_status()
    data = resp.json()
    items = data.get("items", data) if isinstance(data, dict) else data
    if not isinstance(items, list):
        items = []
    return [r for r in items if (r.get("timestamp_created") or "") >= since_iso]


def _reply_field(reply: dict, *candidates: str, default: str = "") -> str:
    for key in candidates:
        if reply.get(key):
            return reply[key]
    return default


def classify_and_suggest(reply: dict, client: anthropic.Anthropic) -> tuple[str, str]:
    body = _reply_field(reply, "body", "text", "email_body")[:2000]
    subject = _reply_field(reply, "subject", "email_subject")
    sender = _reply_field(reply, "from_address", "reply_from", "sender_email")

    prompt = f"""You are classifying a cold-email reply for a UK B2B AI company.

Classify into EXACTLY one of: INTERESTED, BOOK_A_CALL, OBJECTION, OOO, UNSUBSCRIBE, NOT_INTERESTED, WRONG_PERSON, QUESTION

Rules:
- INTERESTED: positive, wants to learn more but hasn't asked to book
- BOOK_A_CALL: explicitly asks to schedule / meet
- OBJECTION: engaged pushback worth addressing (e.g. "we already use X", "not the right time but...")
- OOO: out-of-office auto-reply
- UNSUBSCRIBE: any opt-out / remove-me request
- NOT_INTERESTED: flat rejection, no engagement
- WRONG_PERSON: forwarded or told to contact someone else
- QUESTION: asking something specific before deciding

Subject: {subject}
From: {sender}
Body:
{body}

Reply with ONLY valid JSON — no markdown, no explanation:
{{"class": "<ONE_OF_THE_ABOVE>", "suggested": "<reply or null>"}}

For "suggested": if class is INTERESTED, BOOK_A_CALL, or OBJECTION write a 2–4 sentence warm UK-English reply that references something specific they said. For INTERESTED and BOOK_A_CALL include {CAL_LINK} naturally. Sign off "Tom — Voqal AI". For all other classes set "suggested" to null."""

    msg = client.messages.create(
        model="claude-haiku-4-5-20251001",
        max_tokens=600,
        messages=[{"role": "user", "content": prompt}],
    )
    raw = msg.content[0].text.strip()
    parsed = json.loads(raw)
    cls = parsed.get("class", "QUESTION")
    if cls not in ALL_CLASSES:
        cls = "QUESTION"
    return cls, parsed.get("suggested") or ""


def delete_lead(email: str, campaign_id: str, api_key: str) -> None:
    try:
        requests.delete(
            f"{INSTANTLY_BASE}/leads/{email}",
            headers={"Authorization": f"Bearer {api_key}"},
            params={"campaign_id": campaign_id},
            timeout=10,
        )
    except Exception:
        pass  # best-effort; no notification for unsubscribe failures


def _send_telegram_payload(payload: dict, bot_token: str) -> bool:
    url = f"{TELEGRAM_BASE}/bot{bot_token}/sendMessage"
    resp = requests.post(url, json=payload, timeout=15)
    return resp.json().get("ok", False)


def send_telegram(text: str, bot_token: str, chat_id: int) -> None:
    payload = {"chat_id": chat_id, "text": text, "parse_mode": "HTML"}
    if not _send_telegram_payload(payload, bot_token):
        # Retry once without HTML parse_mode
        payload.pop("parse_mode")
        _send_telegram_payload(payload, bot_token)


def build_hot_message(cls: str, reply: dict, campaign_label: str, suggested: str) -> str:
    sender_name = escape(_reply_field(reply, "from_name", "sender_name"), quote=False)
    sender_email = escape(_reply_field(reply, "from_address", "reply_from", "sender_email"), quote=False)
    subject = escape(_reply_field(reply, "subject", "email_subject"), quote=False)
    body = escape(_reply_field(reply, "body", "text", "email_body")[:600], quote=False)

    icon = "🔥" if cls in ("INTERESTED", "BOOK_A_CALL") else "📌"
    lines = [
        f'{icon} <b>{cls}</b> — {sender_name} ({sender_email})',
        f"Campaign: {campaign_label} • Subject: {subject}",
        "",
        f'"{body}"',
    ]
    if suggested:
        lines += ["", "💡 <b>Suggested response:</b>", escape(suggested, quote=False)]
    return "\n".join(lines)


def main() -> None:
    api_key = get_env("INSTANTLY_API_KEY")
    bot_token = get_env("TELEGRAM_BOT_TOKEN")
    chat_id = int(get_env("TELEGRAM_CHAT_ID"))
    uk_id = get_env("UK_CAMPAIGN_ID")
    us_id = get_env("US_CAMPAIGN_ID")

    since_iso = (datetime.now(timezone.utc) - timedelta(hours=2, minutes=10)).isoformat()

    client = anthropic.Anthropic()

    errors: list[str] = []
    all_replies: list[tuple[str, dict, str]] = []  # (label, reply, campaign_id)

    for label, cid in [("UK", uk_id), ("US", us_id)]:
        try:
            replies = fetch_replies(cid, since_iso, api_key)
            for r in replies:
                all_replies.append((label, r, cid))
        except Exception as exc:
            errors.append(f"{label}: {exc}")

    if len(errors) == 2:
        send_telegram(
            f"⚠️ Hot-reply watcher couldn't reach Instantly: {'; '.join(errors)}",
            bot_token,
            chat_id,
        )
        sys.exit(1)

    hot: list[tuple[str, dict, str, str]] = []

    for label, reply, cid in all_replies:
        try:
            cls, suggested = classify_and_suggest(reply, client)
        except Exception as exc:
            print(f"Classification error for reply in {label}: {exc}", file=sys.stderr)
            continue

        if cls == "UNSUBSCRIBE":
            email = _reply_field(reply, "from_address", "reply_from", "sender_email")
            if email:
                delete_lead(email, cid, api_key)
            continue

        if cls in HOT_CLASSES:
            hot.append((cls, reply, label, suggested))

    if not hot:
        return  # silent — nothing to report

    for cls, reply, label, suggested in hot[:5]:
        msg = build_hot_message(cls, reply, label, suggested)
        send_telegram(msg, bot_token, chat_id)
        time.sleep(0.3)

    remainder = len(hot) - 5
    if remainder > 0:
        send_telegram(
            f"+{remainder} more hot replies — see morning brief",
            bot_token,
            chat_id,
        )


if __name__ == "__main__":
    main()

File: netlify/functions/hot-reply-watcher.js
'use strict';

// ─── Helpers ────────────────────────────────────────────────────────────────

function htmlEscape(str) {
  return String(str)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');
}

function trimBody(text, max = 600) {
  if (!text) return '';
  const s = text.trim();
  return s.length <= max ? s : s.slice(0, max - 1) + '…';
}

// ─── Classification ──────────────────────────────────────────────────────────

const PATTERNS = [
  { cls: 'UNSUBSCRIBE',    re: /unsubscribe|opt.?out|remove\s+me|stop\s+email|don'?t\s+email|do\s+not\s+contact/i },
  { cls: 'OOO',            re: /out\s+of\s+(the\s+)?office|ooo\b|on\s+(annual\s+)?leave|on\s+holiday|on\s+vacation|away\s+until|returning\s+on|back\s+on|back\s+in\s+the\s+office/i },
  { cls: 'BOOK_A_CALL',    re: /book\s+(a\s+)?(call|meeting|time|demo)|schedule\s+(a\s+)?(call|meeting|demo)|let'?s\s+(hop\s+on|chat|catch\s+up|connect|meet)|happy\s+to\s+(chat|talk|connect)|set\s+up\s+(a\s+)?(call|meeting|time)|calendar|availability|when\s+are\s+you\s+(free|available)/i },
  { cls: 'INTERESTED',     re: /interested|tell\s+me\s+more|sounds\s+good|sounds\s+great|love\s+to\s+(learn|hear|know)|learn\s+more|how\s+does\s+it\s+work|what\s+does\s+it\s+(do|cost)|yes\s+please|keen\s+to|open\s+to\s+a|would\s+love/i },
  { cls: 'NOT_INTERESTED', re: /not\s+interested|no\s+thanks|no\s+thank\s+you|not\s+for\s+us|not\s+looking|not\s+right\s+now|not\s+at\s+this\s+(time|stage)|pass\s+on\s+this|not\s+in\s+(the\s+)?market/i },
  { cls: 'WRONG_PERSON',   re: /wrong\s+person|wrong\s+(email|address)|not\s+the\s+right|don'?t\s+handle|not\s+(my\s+)?responsible|not\s+my\s+(area|department|remit)/i },
];

function classify(body) {
  const text = body || '';
  for (const { cls, re } of PATTERNS) {
    if (re.test(text)) return cls;
  }
  // Default heuristic: question → QUESTION, hesitation → OBJECTION
  if (/\?/.test(text) || /^(how|what|when|where|who|why|can|could|do|are|is|will)\b/im.test(text)) {
    return 'QUESTION';
  }
  if (/\b(but|however|although|concern|worried|not\s+sure|budget|cost|price|contract|commitment|currently|already\s+have)\b/i.test(text)) {
    return 'OBJECTION';
  }
  return 'QUESTION';
}

// An OBJECTION is "engaged" if it also contains a question or affirmative intent
function isEngagedObjection(body) {
  const text = body || '';
  return /\?/.test(text) || /interested|open|tell\s+me|how\s+does|worth|consider/i.test(text);
}

// ─── Suggested response templates ───────────────────────────────────────────

function suggestReply(cls, senderName, subject, body) {
  const first = (senderName || '').split(' ')[0] || 'there';
  const snippet = trimBody(body, 120);

  if (cls === 'BOOK_A_CALL') {
    return (
      `Thanks so much for getting back to me, ${first} — brilliant to hear from you! ` +
      `I'd love to get something in the diary. You can grab a time that suits you here: https://cal.com/voqalai/discovery ` +
      `— it's just a relaxed 20-minute discovery call, no pressure at all. ` +
      `Looking forward to speaking soon. Tom — Voqal AI`
    );
  }
  if (cls === 'INTERESTED') {
    return (
      `Great to hear from you, ${first}, and really pleased to know you're keen to find out more! ` +
      `The best next step is a quick discovery call so I can understand your situation properly and show you what's possible — you can book a time that works for you here: https://cal.com/voqalai/discovery ` +
      `It's only 20 minutes and completely no-obligation. ` +
      `Tom — Voqal AI`
    );
  }
  if (cls === 'OBJECTION') {
    const ref = snippet ? ` You mentioned "${snippet}" — that's something I'm happy to address properly rather than brush over.` : '';
    return (
      `Thanks for the honest reply, ${first} — I really appreciate it.${ref} ` +
      `Would it be worth a quick 20-minute call so I can give you a straight answer? ` +
      `Tom — Voqal AI`
    );
  }
  return `Thanks for getting back to me, ${first}. Tom — Voqal AI`;
}

// ─── Telegram ────────────────────────────────────────────────────────────────

async function telegramSend(token, chatId, text) {
  const payload = JSON.stringify({ chat_id: chatId, text, parse_mode: 'HTML' });
  const url = `https://api.telegram.org/bot${token}/sendMessage`;

  let res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: payload,
  });
  let data = await res.json();

  if (!data.ok) {
    // Retry once without parse_mode
    const fallback = JSON.stringify({ chat_id: chatId, text: text.replace(/<[^>]+>/g, '') });
    res = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: fallback,
    });
    data = await res.json();
  }

  return data;
}

// ─── Instantly ───────────────────────────────────────────────────────────────

async function fetchReplies(campaignId, apiKey) {
  const url = `https://api.instantly.ai/api/v2/emails/reply?campaign_id=${encodeURIComponent(campaignId)}&limit=100`;
  const res = await fetch(url, {
    headers: { Authorization: `Bearer ${apiKey}` },
  });
  if (!res.ok) {
    throw new Error(`HTTP ${res.status} for campaign ${campaignId}`);
  }
  const data = await res.json();
  // v2 returns { items: [...] } or an array directly
  return Array.isArray(data) ? data : (data.items || data.data || []);
}

async function deleteLead(email, campaignId, apiKey) {
  const url = `https://api.instantly.ai/api/v2/leads/${encodeURIComponent(email)}?campaign_id=${encodeURIComponent(campaignId)}`;
  try {
    await fetch(url, {
      method: 'DELETE',
      headers: { Authorization: `Bearer ${apiKey}` },
    });
  } catch (_) {
    // Best-effort; don't throw
  }
}

// ─── Field normaliser ────────────────────────────────────────────────────────

function normaliseReply(item) {
  return {
    timestamp: item.timestamp_created || item.created_at || item.timestamp || '',
    senderEmail: item.sender_email || item.from_address || item.from || '',
    senderName:  item.sender_name  || item.from_name   || item.from_full_name || '',
    subject:     item.subject      || item.email_subject || '(no subject)',
    body:        item.body         || item.reply_body   || item.email_body    || '',
  };
}

// ─── Main handler ─────────────────────────────────────────────────────────────

exports.handler = async function handler() {
  const {
    INSTANTLY_API_KEY,
    TELEGRAM_BOT_TOKEN,
    TELEGRAM_CHAT_ID,
    UK_CAMPAIGN_ID,
    US_CAMPAIGN_ID,
  } = process.env;

  const chatId = parseInt(TELEGRAM_CHAT_ID, 10);

  // since_iso = now - 2h10m
  const sinceMs = Date.now() - (2 * 60 + 10) * 60 * 1000;
  const sinceIso = new Date(sinceMs).toISOString();

  // ── Fetch from both campaigns ──────────────────────────────────────────────
  const campaigns = [
    { id: UK_CAMPAIGN_ID, label: 'UK' },
    { id: US_CAMPAIGN_ID, label: 'US' },
  ];

  let allErrors = [];
  let allResults = []; // { campaign, reply }

  for (const { id, label } of campaigns) {
    try {
      const items = await fetchReplies(id, INSTANTLY_API_KEY);
      const recent = items.filter(item => {
        const ts = item.timestamp_created || item.created_at || item.timestamp || '';
        return ts >= sinceIso;
      });
      for (const item of recent) {
        allResults.push({ campaign: label, campaignId: id, reply: normaliseReply(item) });
      }
    } catch (err) {
      allErrors.push(`${label}: ${err.message}`);
    }
  }

  // Both campaigns failed → error alert + exit
  if (allErrors.length === campaigns.length) {
    await telegramSend(
      TELEGRAM_BOT_TOKEN,
      chatId,
      `⚠️ Hot-reply watcher couldn't reach Instantly: ${allErrors.join('; ')}`,
    );
    return { statusCode: 200, body: 'error-alert-sent' };
  }

  // ── Classify and partition ──────────────────────────────────────────────────
  const hotReplies = [];

  for (const { campaign, campaignId, reply } of allResults) {
    const cls = classify(reply.body);

    if (cls === 'UNSUBSCRIBE') {
      await deleteLead(reply.senderEmail, campaignId, INSTANTLY_API_KEY);
      continue;
    }

    const isHot =
      cls === 'INTERESTED' ||
      cls === 'BOOK_A_CALL' ||
      (cls === 'OBJECTION' && isEngagedObjection(reply.body));

    if (isHot) {
      hotReplies.push({ cls, campaign, reply });
    }
  }

  // ── Nothing hot → stay silent ───────────────────────────────────────────────
  if (hotReplies.length === 0) {
    return { statusCode: 200, body: 'silent' };
  }

  // ── Send up to 5 individual messages ───────────────────────────────────────
  const toSend   = hotReplies.slice(0, 5);
  const overflow = hotReplies.length - toSend.length;

  for (const { cls, campaign, reply } of toSend) {
    const name    = htmlEscape(reply.senderName  || reply.senderEmail);
    const email   = htmlEscape(reply.senderEmail);
    const subject = htmlEscape(reply.subject);
    const body    = htmlEscape(trimBody(reply.body, 600));
    const suggest = htmlEscape(suggestReply(cls, reply.senderName, reply.subject, reply.body));

    const text =
      `🔥 <b>${cls}</b> — ${name} (${email})\n` +
      `Campaign: ${campaign} • Subject: ${subject}\n\n` +
      `"${body}"\n\n` +
      `💡 <b>Suggested response:</b>\n${suggest}`;

    await telegramSend(TELEGRAM_BOT_TOKEN, chatId, text);
  }

  if (overflow > 0) {
    await telegramSend(
      TELEGRAM_BOT_TOKEN,
      chatId,
      `+${overflow} more hot ${overflow === 1 ? 'reply' : 'replies'} — see morning brief`,
    );
  }

  return { statusCode: 200, body: `sent:${toSend.length}` };
};
